x-pack/plugin/ql/src/main/java/org/elasticsearch/xpack/ql/expression/AttributeMap.java

```diff
@@ -258,10 +260,21 @@ public E get(Object key) {
 
     @Override
     public E getOrDefault(Object key, E defaultValue) {
-        E e;
-        return (((e = get(key)) != null) || containsKey(key))
-            ? e
-            : defaultValue;
+        if (key instanceof NamedExpression) {
+            return delegate.getOrDefault(new AttributeWrapper(((NamedExpression) key).toAttribute()), defaultValue);
+        }
+        return defaultValue;
+    }
+
+    @SuppressWarnings({ "unchecked" })
+    public E resolve(Object key, E defaultValue) {
+        AttributeMap<Object> map = (AttributeMap<Object>) this;
+        Object candidate = defaultValue;
+        Object value = null;
+        while ((value = map.getOrDefault(key, NOT_FOUND)) != NOT_FOUND && value != key) {
+            key = candidate = value;
+        }
+        return (E) candidate;
     }
```

Comment:
As this method builds on other types methods in this class, it shouldn't have unchecked warnings nor used `NOT_FOUND`. Also not sure why the `candidate` is returned instead of `value`, should be the other way around.

Further more this implementation looks buggy: if the map contains `(e,e)` calling resolve(e, default) would incorrectly return `default` and not `e`.
Further more there's no handling of cycles which leads to an infinite loop: (a, b), (b, a).

Below a suggestion to fix both issues:

```suggestion
    public E resolveOrDefault(Object key, E defaultValue) {
        E value = defaultValue;
        E candidate;
        int allowedLookups = 10;
        while (key != value && ((candidate = get(key)) != null || containsKey(key))) {
            if (--allowedLookups == 0) {
	           throw new QlIllegalArgumentException("Potential cycle detected");
	        }

            key = candidate;
            value = candidate;
        }
        return value;
    }
```

Reply:
Good catch with the default value. I was thinking of adding the limitation on the number of lookups, but seemed we would need quite a huge bug in our AttributeMap population for it to happen with more than one hops in the cycle. Anyways, I added this defensive check.

The suggested code won't really work, because:

1.`resolveOrDefault(e, e)` will always return `e` even if the map has `e -> f` mapping.
2. The loop will always terminate after the first lookup (`key == value == candidate`)

Anyways, the updated code fixes the bug you found.